README.md
## Xcode Coverage for Commit Reporter (xcccr)
This dead simple tool only works with Apple's `xccov` generated code coverage reports in json format. It takes two reports, diffs and spits out warnings and errors in GitHub Actions compatible format.

### CLI usage
 Flag   | Description
--------|------------
 `-cur` | For path of "current report" file. A report to be checked.
 `-lst` | For path of "last report" file. A report to be base for determining whether the coverage metrics is lowered or became higher in "current" report.
 `-tol` | Percent to be tolerated before error. E.g., last is 70%, current is 65% and tol is 10 means no error.
 `-proj` | Project root to get relative paths. If empty and missing from config working dir would be used. Usually in GitHub Actions you just want the working dir.
 `-nw` | Comma separated strings of paths not to warn. When a path has one of the values of this list warnings for the files on path would be muted. Coverage would still be metered.
 `-rg` | Regex filter, pass it in quotes e.g. `-rg="[\s\S]\.swift"`. Matching paths would be ignored. When used with `-i` the result is opposite: only matching paths would be visible to reporter. Please note that if regex is used targets and paths filtering set in config has no effect.
 `-i` | Invert filter. Applied to -rg and to targets/paths filter lists in config. If no filters supplied prevents reporting allowing nothing.
 `-si` | Selectively invert filters. When `-i` is set it is equivalent to `-si=prtw` and always have precedence over this flag. Values here are: `p` — inverts paths filter, `r` — inverts regex, `t` — inverts targets filter, `w` — inverts warnings filter.
 `-m` | Include masked files coverage in total metrics. If true warnings would be produced only for unfiltered targets/paths, but total coverage would be reported for all the files as listed in original xccov json.
 `-loc` | Count covered LOCs diff instead of percent coverage.
 `-z` | Don't error on 0% coverage, produce warning only.
 `-cfg` | Path to config. Parameters passed with the flags have precedence and overrides the values stored in config. Config passed with this flag overrides default config at `.xcccr.toml`. If there aren't any config default values are used.


 If there aren't last report or it isn't needed it's ok to pass only the current one. Current report also may be piped in to xcccr like that `cat report.json | ./xcccr`  

 ### Config values

Please check `.xcccr.toml.example` for reference.


RELEASENOTES.md

### Hotfix (0.3.3)
- Falls back now as intended to current report when last report file is missing. A `no such file or directory` error wasn't overriden correctly.
- Fixed some typos in human readable texts (e.g. readme and cli help).

### Hotfix (0.3.2)
- Added missing warning message for file coverage % when reduced.
- Some warnings weren't visible. Seem GitHub Actions warnings should be oneliners. Removed newline symbols.
- Some code duplications cleaned.

### Hotfix (0.3.1)
- "Increased" and "Decreased" messages was swapped. 

### Fixed (0.3.0)
- File paths for warnings was always relative to working dir only. `ProjectPath` setting in config and `proj` flag was completely ignored.
- Removed some code duplication.
- Data race is cured by adding a wait group to wait for all file reports being processed. Sometimes looping through all the files was ended prior the last report is collected from channel leading to incomplete counts. No more.

### Changed (0.3.0)
- Now warnings for functions coverage aren't shown per line, but instead are included in file coverage warnings. It still includes xccov messages with the line numbers.
- New filter added to mute warnings for paths without excluding the path from coverage report.
- Filters may be inverted individually with new flags and config settings. Please check readme and config example for usage.
- No breaking changes were introduced.

main.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"os"
	"strings"
	"time"
)

func init() {
	rand.Seed(time.Now().UnixNano())
}

func main() {
	fmt.Println("Xcode Coverage for Commit Reporter (https://github.com/illabo/xcccr)")

	rnCnd, err := prepareRunConditions()
	if err != nil {
		log.Fatal(err)
	}

	performReport(rnCnd)
}

func getWorkdir() string {
	dir, _ := os.Getwd()
	if dir == "" {
		dir, _ = os.UserHomeDir()
	}
	if dir == "" {
		dir = "~/"
	}
	return appendSlash(dir)
}

func appendSlash(pathStr string) string {
	if strings.HasSuffix(pathStr, "/") {
		return pathStr
	}
	return pathStr + "/"
}

func randomStr(ln int) string {
	set := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	rb := make([]byte, ln)
	for i := range rb {
		rb[i] = set[rand.Intn(len(set))]
	}
	return string(rb)
}


types.go
package main

import (
	"regexp"
	"sync"
)

// Report
type XCCoverageReport struct {
	CoveredLines    int      `json:"coveredLines"`
	LineCoverage    float64  `json:"lineCoverage"`
	Targets         []Target `json:"targets"`
	ExecutableLines int      `json:"executableLines"`
}
type Function struct {
	CoveredLines    int     `json:"coveredLines"`
	LineCoverage    float64 `json:"lineCoverage"`
	LineNumber      int     `json:"lineNumber"`
	ExecutionCount  int     `json:"executionCount"`
	Name            string  `json:"name"`
	ExecutableLines int     `json:"executableLines"`
}
type File struct {
	CoveredLines    int        `json:"coveredLines"`
	LineCoverage    float64    `json:"lineCoverage"`
	Path            string     `json:"path"`
	Functions       []Function `json:"functions"`
	Name            string     `json:"name"`
	ExecutableLines int        `json:"executableLines"`
}
type Target struct {
	CoveredLines     int     `json:"coveredLines"`
	LineCoverage     float64 `json:"lineCoverage"`
	Files            []File  `json:"files"`
	Name             string  `json:"name"`
	ExecutableLines  int     `json:"executableLines"`
	BuildProductPath string  `json:"buildProductPath"`
}

// App config
type Config struct {
	ProjectPath          string
	FilterTargets        []string
	FilterPaths          []string
	FilterWarnPaths      []string
	FilterPattern        string
	InvertFilter         bool
	InvertTargetFilter   bool
	InvertPathFilter     bool
	InvertRegexp         bool
	InvertWarnpathFilter bool
	IncludeMasked        bool
	ZeroWarnOnly         bool
	MeterLOC             bool
	Tolerance            int
}
type RunConditions struct {
	LastReport    *XCCoverageReport
	CurrentReport *XCCoverageReport
	Config        *Config
	Regexp        *regexp.Regexp
	DiffUnitChan  chan DiffUnit
	TotalUnitChan chan DiffUnit
	WaitGroup     *sync.WaitGroup
}

// Coverage diff unit
type DiffUnit struct {
	LastExecutableLines    int
	LastCoveredLines       int
	CurrentExecutableLines int
	CurrentCoveredLines    int
}


reporter.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func performReport(rnCnd *RunConditions) {
	last := rnCnd.LastReport
	current := rnCnd.CurrentReport
	var totalsUnit DiffUnit

	go func(tu *DiffUnit) {
		for {
			totalsUnit = <-rnCnd.TotalUnitChan
			rnCnd.WaitGroup.Done()
		}
	}(&totalsUnit)
	go unitCounter(rnCnd)

	lastFlatTgs := flattenTargets(last)
	for _, ct := range current.Targets {
		if targetIsAllowed(rnCnd, &ct) {
			if lt, ok := lastFlatTgs[ct.Name]; ok {
				diffTargets(rnCnd, &lt, &ct)
			} else {
				diffTargets(rnCnd, &Target{}, &ct)
			}
		}
	}

	rnCnd.WaitGroup.Wait()

	produceVerdict(rnCnd.Config, &totalsUnit, last, current)
}

func produceVerdict(cfg *Config, tu *DiffUnit, last, current *XCCoverageReport) {
	if cfg.IncludeMasked {
		tu.LastExecutableLines = last.ExecutableLines
		tu.LastCoveredLines = last.CoveredLines
		tu.CurrentExecutableLines = current.ExecutableLines
		tu.CurrentCoveredLines = current.CoveredLines
	}
	lastPct := percentify(linesRatio(tu.LastCoveredLines, tu.LastExecutableLines))
	currentPct := percentify(linesRatio(tu.CurrentCoveredLines, tu.CurrentExecutableLines))
	covMsg := "Code coverage is %s"

	if lastPct > currentPct {
		covMsg = fmt.Sprintf(covMsg, "decreased")
	} else if lastPct < currentPct {
		covMsg = fmt.Sprintf(covMsg, "increased")
	} else {
		covMsg = fmt.Sprintf(covMsg, "not changed")
	}
	var curCovMsg, lstCovMsg string
	if cfg.MeterLOC {
		curCovMsg = fmt.Sprintf("currently covered %dLOC", tu.CurrentCoveredLines)
		lstCovMsg = fmt.Sprintf("last coverage was %dLOC", tu.LastCoveredLines)
	} else {
		curCovMsg = fmt.Sprintf("currently covered %d%%", currentPct)
		lstCovMsg = fmt.Sprintf("last coverage was %d%%", lastPct)
	}
	if len(last.Targets) == 0 {
		lstCovMsg = "last coverage is unavailable"
	}
	covMsg = fmt.Sprintf("%s: %s, %s.", covMsg, curCovMsg, lstCovMsg)
	if lastPct > currentPct+cfg.Tolerance {
		emmitError(covMsg)
	}
	currentPctZero := currentPct == 0
	if currentPctZero && cfg.ZeroWarnOnly == false {
		emmitError(covMsg)
	}
	if currentPctZero {
		covMsg = fmt.Sprintf("::warning::%s", covMsg)
	}
	exitMessage(covMsg)
}

func emmitError(message string) {
	fmt.Printf("::error::%s\n", message)
	os.Exit(1)
}

func exitMessage(message string) {
	fmt.Println(message)
	os.Exit(0)
}

func diffTargets(rnCnd *RunConditions, lastTgt, currentTgt *Target) {
	ltgtFlatFiles := flattenFiles(lastTgt)
	// Wait for every file coverage to be reported.
	rnCnd.WaitGroup.Add(len(currentTgt.Files))
	for _, curFile := range currentTgt.Files {
		if pathIsAllowed(rnCnd, &curFile) == false {
			// Path is excluded from coverage report.
			// Don't expect coverage unit. Try next.
			rnCnd.WaitGroup.Done()
			continue
		}

		var lfsFlatFuncs map[string]Function
		lstFile := ltgtFlatFiles[curFile.Name]
		curFilePth := strings.TrimPrefix(curFile.Path, rnCnd.Config.ProjectPath)

		lfsFlatFuncs = flattenFuncs(&lstFile)
		rnCnd.DiffUnitChan <- DiffUnit{
			LastExecutableLines:    lstFile.ExecutableLines,
			LastCoveredLines:       lstFile.CoveredLines,
			CurrentExecutableLines: curFile.ExecutableLines,
			CurrentCoveredLines:    curFile.CoveredLines,
		}

		if warnIsAllowed(rnCnd, &curFile) == false {
			// Don't warn, count coverage only.
			continue
		}

		var fileReport string
		if curFile.CoveredLines < lstFile.CoveredLines && rnCnd.Config.MeterLOC {
			fileReport = fmt.Sprintf(
				"File coverage is reduced: %s",
				fmt.Sprintf(
					"covered %dLOC, was %dLOC.",
					curFile.CoveredLines,
					lstFile.CoveredLines,
				))
		}
		curPct := percentify(linesRatio(curFile.CoveredLines, curFile.ExecutableLines))
		lstPct := percentify(linesRatio(lstFile.CoveredLines, lstFile.ExecutableLines))
		if curPct < lstPct && !rnCnd.Config.MeterLOC {
			fileReport = fmt.Sprintf(
				"File coverage is reduced: %s",
				fmt.Sprintf("covered %d%%, was %d%%.",
					curPct,
					lstPct,
				))
		}
		if curFile.CoveredLines == 0 {
			fileReport = "File is not covered."
		}

		for _, curFunc := range curFile.Functions {
			lstFunc := lfsFlatFuncs[curFunc.Name]
			cov := warnFuncCov(&lstFunc, &curFunc)
			if cov != "" {
				fileReport = fmt.Sprintf("%s %s", fileReport, cov)
			}
		}
		if strings.TrimSpace(fileReport) != "" {
			fmt.Printf("::warning file=%s::%s\n", curFilePth, fileReport)
		}

	}
}

func warnFuncCov(lstFunc, curFunc *Function) string {
	if lstFunc != nil && lstFunc.LineCoverage > curFunc.LineCoverage {
		return fmt.Sprintf(
			"Line %d: %s coverage is lowered to %d%% (was %d%%).",
			curFunc.LineNumber,
			curFunc.Name,
			percentify(curFunc.LineCoverage),
			percentify(lstFunc.LineCoverage),
		)

	}
	return warnZeroFuncCov(curFunc)
}

func warnZeroFuncCov(curFunc *Function) string {
	if percentify(curFunc.LineCoverage) == 0 {
		return fmt.Sprintf(
			"Line %d: %s is not covered.",
			curFunc.LineNumber,
			curFunc.Name,
		)
	}
	return ""
}

func linesRatio(cov, exec int) float64 {
	if exec == 0 {
		return 0
	}
	return float64(cov) / float64(exec)
}

func percentify(linecov float64) int {
	return int(linecov * 100)
}

func flattenTargets(covR *XCCoverageReport) map[string]Target {
	flatTargets := map[string]Target{}
	for _, t := range covR.Targets {
		flatTargets[t.Name] = t
	}
	return flatTargets
}

func flattenFiles(tgtR *Target) map[string]File {
	flatFiles := map[string]File{}
	for _, f := range tgtR.Files {
		flatFiles[f.Name] = f
	}
	return flatFiles
}

func flattenFuncs(filR *File) map[string]Function {
	flatFuncs := map[string]Function{}
	for _, f := range filR.Functions {
		flatFuncs[f.Name] = f
	}
	return flatFuncs
}

func targetIsAllowed(rnCnd *RunConditions, tgt *Target) bool {
	if rnCnd.Regexp != nil {
		return true // Allow any target and let regex sortout filtering on paths.
	}
	return elSubInList(rnCnd.Config.FilterTargets, tgt.Name) == rnCnd.Config.InvertTargetFilter
}

func pathIsAllowed(rnCnd *RunConditions, fl *File) bool {
	if rnCnd.Regexp != nil {
		return elemAllowedWithRegex(rnCnd, fl.Path)
	}
	return elSubInList(rnCnd.Config.FilterPaths, fl.Path) == rnCnd.Config.InvertPathFilter
}

func warnIsAllowed(rnCnd *RunConditions, fl *File) bool {
	return elSubInList(rnCnd.Config.FilterWarnPaths, fl.Path) == rnCnd.Config.InvertWarnpathFilter
}

func elSubInList(elList []string, elName string) bool {
	for _, el := range elList {
		if strings.Contains(elName, el) {
			return true
		}
	}
	return false
}

func elemAllowedWithRegex(rnCnd *RunConditions, el string) bool {
	if rnCnd.Regexp == nil {
		return true
	}
	return (rnCnd.Regexp.Find([]byte(el)) != nil) == rnCnd.Config.InvertRegexp
}

func unitCounter(rnCnd *RunConditions) {
	totalDu := DiffUnit{}
	for {
		u := <-rnCnd.DiffUnitChan
		totalDu.CurrentCoveredLines += u.CurrentCoveredLines
		totalDu.CurrentExecutableLines += u.CurrentExecutableLines
		totalDu.LastCoveredLines += u.LastCoveredLines
		totalDu.LastExecutableLines += u.LastExecutableLines

		rnCnd.TotalUnitChan <- totalDu
	}
}


configurator.go
package main

import (
	"bufio"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"

	"github.com/BurntSushi/toml"
)

func readConfig(path string) (*Config, error) {
	var cfg Config
	_, err := toml.DecodeFile(path, &cfg)
	return &cfg, err
}

func prepareRunConditions() (rnCnd *RunConditions, err error) {
	rnCnd = &RunConditions{}

	dumbValue := dumbValue()

	lstPth := flag.String(
		"lst",
		"",
		"Pass the path of last report to compare current report to.",
	)
	curPth := flag.String(
		"cur",
		"",
		"Pass the path of current report.",
	)
	tolerance := flag.String("tol",
		dumbValue,
		"Percent to be tolerated before error. E.g. last is 70%, current is 65% and tol is 10 means no error.",
	)
	projRoot := flag.String(
		"proj",
		getWorkdir(),
		"Project root to get relative paths. If empty and missing from config working dir would be used.",
	)
	mute := flag.String(
		"nw",
		"",
		"Comma separated strings of paths not to warn. When a path has one of the values of this list warnings for the files on path would be muted. Coverage would still be metered.",
	)
	filtr := flag.String("rg",
		dumbValue,
		"Regex filter. Applied to paths.",
	)
	fInvert := flag.Bool(
		"i",
		false,
		"Invert filter. Applied to -rg and to targets/paths filter lists in config. If no filters supplied prevents reporting allowing nothing.",
	)
	selInvert := flag.String(
		"si",
		"",
		"Selectively invert filters. When `-i` is set it is equivalent to `-si=prtw` and always have precedence over this flag. Values here are:\n`p` — inverts paths filter,\n`r` — inverts regex,\n`t` — inverts targets filter,\n`w` — inverts warnings filter.",
	)
	includeMasked := flag.Bool(
		"m",
		false,
		"Include masked files coverage in total metrics. If true warnings would be produced only for unfiltered targets/paths, but total coverage would be calculated for all the files.",
	)
	meterLOC := flag.Bool(
		"loc",
		false,
		"Count covered LOCs diff instead of percent coverage.",
	)
	zeroWarn := flag.Bool(
		"z",
		false,
		"Don't error on 0%% coverage, produce warning only.",
	)
	cfgPth := flag.String(
		"cfg",
		"",
		"Path to config.",
	)

	flag.Parse()

	if *cfgPth == "" {
		*cfgPth = ".xcccr.toml"
	}
	cfg, err := readConfig(*cfgPth)
	if err != nil && !os.IsNotExist(err) {
		return
	}
	if *tolerance != dumbValue {
		i, e := strconv.Atoi(*tolerance)
		if i < 0 || i > 100 {
			e = errors.New("out of allowed range")
		}
		if e != nil {
			err = fmt.Errorf("tolerance -tol must be an int in range 0 to 100. %w", err)
			return
		}
		cfg.Tolerance = i
	}
	if *mute != "" {
		cfg.FilterWarnPaths = strings.Split(*mute, ",")
	}
	if *filtr != dumbValue {
		cfg.FilterPattern = *filtr
	}
	if *fInvert {
		cfg.InvertFilter = true
	}
	if cfg.InvertFilter {
		cfg.InvertPathFilter = true
		cfg.InvertRegexp = true
		cfg.InvertTargetFilter = true
		cfg.InvertWarnpathFilter = true
	}
	for _, c := range *selInvert {
		if c == 'p' {
			cfg.InvertPathFilter = true
		}
		if c == 'r' {
			cfg.InvertRegexp = true
		}
		if c == 't' {
			cfg.InvertTargetFilter = true
		}
		if c == 'w' {
			cfg.InvertWarnpathFilter = true
		}
	}
	if *includeMasked {
		cfg.IncludeMasked = true
	}
	if *meterLOC {
		cfg.MeterLOC = true
	}
	if *zeroWarn {
		cfg.ZeroWarnOnly = true
	}
	cfg.ProjectPath = *projRoot
	if cfg.FilterPattern != "" {
		re, e := regexp.Compile(cfg.FilterPattern)
		if e != nil {
			err = fmt.Errorf("filter pattern (aka regexp) is invalid: %w", e)
			return
		}
		rnCnd.Regexp = re
	}

	rnCnd.Config = cfg
	rnCnd.DiffUnitChan = make(chan DiffUnit)
	rnCnd.TotalUnitChan = make(chan DiffUnit)

	curXccRep, err := getCurrentCoverage(curPth)
	if err != nil {
		return
	}
	rnCnd.CurrentReport = curXccRep

	lstXccRep, err := getLastCoverage(lstPth)
	if err != nil {
		return
	}
	rnCnd.LastReport = lstXccRep

	rnCnd.WaitGroup = &sync.WaitGroup{}

	return
}

func getCurrentCoverage(curPth *string) (curXccRep *XCCoverageReport, err error) {
	var pipeIn []byte
	info, _ := os.Stdin.Stat()
	if info.Mode()&os.ModeNamedPipe != 0 {
		reader := bufio.NewReader(os.Stdin)
		for {
			ln, _, err := reader.ReadLine()
			if err != nil && err == io.EOF {
				break
			}
			pipeIn = append(pipeIn, ln...)
		}
	}

	if *curPth != "" {
		curXccRep, err = readReportFile(*curPth)
		if err != nil {
			return
		}
	}
	// piped in file content always overrides passed with a 'cur' flag
	if len(pipeIn) > 0 {
		if err = json.Unmarshal(pipeIn, &curXccRep); err != nil {
			return
		}
	}
	if *curPth == "" && len(pipeIn) == 0 {
		log.Fatal("No data passed to programme to report coverage.")
	}
	return
}

func getLastCoverage(lstPth *string) (lstXccRep *XCCoverageReport, err error) {
	lstXccRep = &XCCoverageReport{}
	if *lstPth != "" {
		lstXccRep, err = readReportFile(*lstPth)
		notExist := os.IsNotExist(err)
		if err != nil && notExist == false {
			return
		}
		if notExist {
			err = nil
			fmt.Println("Previous report file not found. Continuing with current report only.")
		}
	}
	return
}

func readReportFile(filePath string) (report *XCCoverageReport, err error) {
	report = &XCCoverageReport{}
	filePath = fmt.Sprintf("%s%s", getWorkdir(), filePath)
	_, err = os.Stat(filePath)
	if err != nil {
		return
	}
	fByt, err := ioutil.ReadFile(filePath)
	if err != nil {
		return
	}
	err = json.Unmarshal(fByt, report)

	return
}

func dumbValue() string {
	return randomStr(12)
}
